src/adapter/src/catalog/migrate.rs

```diff
@@ -83,3 +90,42 @@ pub(crate) async fn migrate<S: Append>(catalog: &mut Catalog<S>) -> Result<(), a
 // ****************************************************************************
 // Semantic migrations -- Weird migrations that require access to the catalog
 // ****************************************************************************
+
+// Rewrites all subsource references to be qualified by their IDs, which is the
+// mechanism by which `DeferredObjectName` differentiates between user input and
+// created objects.
+fn deferred_object_name_rewrite(
+    cat: &ConnCatalog,
+    stmt: &mut mz_sql::ast::Statement<Raw>,
+) -> Result<(), anyhow::Error> {
+    if let Statement::CreateSource(CreateSourceStatement {
```

Comment:
Since we're aiming for a general mechanism of derferred names can we use an AST folder here so that simply including a `DeferredObjectName` in your AST will be forced to go through this mechanism?

Reply:
Ah, I was envisioning this as a very targeted migration meant only to catch the existing subsource references that need to get rewritten, but I wanted to make the migration idempotent. Let me know if that doesn't type check for you and glad to make an AST folder.